### PR TITLE
Add LRO Grow-Up

### PR DESCRIPTION
**Pull request by AlexanderSher**
Addresses https://github.com/azure/autorest.csharp/issues/2050 and https://github.com/azure/azure-sdk-for-net/issues/28125

Adds `Convert` helper method for LRO + handwritten methods in `DPGClient` + tests

### CODE REVIEW

File: test/TestServerProjectsLowLevel/dpg-customization/Customization/DPGClient.cs [hunk 1]
```diff
@@ -21,21 +21,11 @@ public partial class DPGClient
         /// <exception cref="ArgumentNullException"> <paramref name="mode"/> is null. </exception>
         public virtual async Task<Response<Product>> GetModelValueAsync(string mode, CancellationToken cancellationToken = default)
         {
-            using var scope = ClientDiagnostics.CreateScope("DPGClient.GetModelValue");
-            scope.Start();
-            try
```
Comment (ShivangiReja) [inline test/TestServerProjectsLowLevel/dpg-customization/Customization/DPGClient.cs:26]: As per the guidance [here](https://github.com/azure/autorest.csharp/issues/2127#issuecomment-1086127240), If the convenience method name is different, then create outer scope and inner scope will be suppressed as we turned on `suppressNestedClientActivities` flag.

>If the grow-up method diverges much from the protocol method (i.e. if they have different names or code in the convenience method could take much time), do open a scope in the convenience API and turn on suppression of inner scopes.

@annelo-msft please correct me if I'm wrong.

Comment (AlexanderSher) [reply test/TestServerProjectsLowLevel/dpg-customization/Customization/DPGClient.cs:26]: That is actually a merge error. Looks like I've picked the wrong file for resolution

Comment (AlexanderSher) [reply test/TestServerProjectsLowLevel/dpg-customization/Customization/DPGClient.cs]: Fixed!

File: test/TestServerProjectsLowLevel/dpg-customization/Customization/DPGClient.cs [hunk 2]
```diff
@@ -101,7 +141,7 @@ public virtual AsyncPageable<Product> GetPagesValuesAsync(string mode, Cancellat
         {
             Argument.AssertNotNull(mode, nameof(mode));
 
-            var context = new RequestContext { CancellationToken = cancellationToken };
+            var context = new RequestContext {CancellationToken = cancellationToken};
```
Comment (annelo-msft) [inline test/TestServerProjectsLowLevel/dpg-customization/Customization/DPGClient.cs:144]: nit: why?

Comment (AlexanderSher) [reply test/TestServerProjectsLowLevel/dpg-customization/Customization/DPGClient.cs:144]: Autoformatting

File: src/assets/Generator.Shared/LowLevelOperationHelpers.cs [hunk 3]
```diff
@@ -62,5 +67,172 @@ public static Operation<Pageable<BinaryData>> ProcessMessage(HttpPipeline pipeli
...
+            private readonly Operation<TFrom> _operation;
+            private readonly ClientDiagnostics _diagnostics;
+            private readonly string _operationName;
+            private readonly string _updateStatusScopeName;
+            private readonly Func<Response, TTo> _convertFunc;
+            private Response<TTo>? _response;
+
+            public ConvertOperation(Operation<TFrom> operation, ClientDiagnostics diagnostics, string operationName, Func<Response, TTo> convertFunc)
+            {
+                _operation = operation;
```
Comment (lirenhe) [inline src/assets/Generator.Shared/LowLevelOperationHelpers.cs:84]: Any parameter check needed in the constructor?

Comment (AlexanderSher) [reply src/assets/Generator.Shared/LowLevelOperationHelpers.cs:84]: No. This is an internal class.

File: test/AutoRest.TestServerLowLevel.Tests/dpg-customization.cs [hunk 4]
```diff
@@ -74,6 +74,61 @@ await foreach (Page<BinaryData> page in allPages.AsPages())
...
+
+            Operation<Product> lro = await new DPGClient(Key, host).LroValueAsync(WaitUntil.Started, "model");
+            Assert.AreEqual(1, diagnosticListener.Scopes.Count);
+            Assert.AreEqual("DPGClient.LroValue", diagnosticListener.Scopes[0].Name);
+            Assert.IsTrue(diagnosticListener.Scopes[0].IsCompleted);
+
+            await lro.WaitForCompletionAsync();
+            Assert.AreEqual(2, diagnosticListener.Scopes.Count);
+            Assert.AreEqual("model", $"{lro.Value.Received}");
+            Assert.AreEqual("DPGClient.LroValue", diagnosticListener.Scopes[1].Name);
```
Comment (annelo-msft) [inline test/AutoRest.TestServerLowLevel.Tests/dpg-customization.cs:91]: The name of the public method call is `WaitForCompletion` ... shouldn't that be the name of the scope?

File: test/AutoRest.TestServerLowLevel.Tests/dpg-customization.cs [hunk 5]
```diff
@@ -74,6 +74,61 @@ await foreach (Page<BinaryData> page in allPages.AsPages())
...
+            using var diagnosticListener = new ClientDiagnosticListener("dpg_customization_LowLevel", asyncLocal: true);
+            CollectionAssert.IsEmpty(diagnosticListener.Scopes);
+
+            Operation<Product> lro = await new DPGClient(Key, host).LroValueAsync(WaitUntil.Started, "model");
+            Assert.AreEqual(1, diagnosticListener.Scopes.Count);
+            Assert.AreEqual("DPGClient.LroValue", diagnosticListener.Scopes[0].Name);
+            Assert.IsTrue(diagnosticListener.Scopes[0].IsCompleted);
+
+            await lro.WaitForCompletionAsync();
+            Assert.AreEqual(2, diagnosticListener.Scopes.Count);
```
Comment (annelo-msft) [inline test/AutoRest.TestServerLowLevel.Tests/dpg-customization.cs:89]: Can we validate that the two scopes are siblings and not nested?

File: test/TestServerProjectsLowLevel/dpg-customization/Customization/DPGClient.cs (hunk 2 above)
Comment (annelo-msft) [reply test/TestServerProjectsLowLevel/dpg-customization/Customization/DPGClient.cs:144]: Could we put it back the other way?  Pretty sure C# style guidelines require spaces here.

File: test/TestServerProjectsLowLevel/dpg-customization/Customization/DPGClient.cs [hunk 6]
```diff
@@ -93,6 +89,50 @@ public virtual Response<Product> PostModel(string mode, Input input, Cancellatio
...
+        /// <exception cref="ArgumentNullException"> <paramref name="mode"/> is null. </exception>
+        public virtual Operation<Product> LroValue(WaitUntil waitUntil, string mode, CancellationToken cancellationToken = default)
+        {
+            var requestContext = new RequestContext { CancellationToken = cancellationToken };
+            using var scope = ClientDiagnostics.CreateScope("DPGClient.LroValue");
+            scope.Start();
+            try
+            {
+                var lro = Lro(waitUntil, mode, requestContext);
+                return LowLevelOperationHelpers.Convert(lro, ClientDiagnostics, "DPGClient.LroValue", r => (Product)r);
```
Comment (annelo-msft) [inline test/TestServerProjectsLowLevel/dpg-customization/Customization/DPGClient.cs:127]: Nit: could we change the parameter ordering to:

`return LowLevelOperationHelpers.Convert(lro, r => (Product)r, ClientDiagnostics, "DPGClient.LroValue");`

This way the main plot comes first and the less important details are later.

**Review by annelo-msft — COMMENTED** [review]
This is really nice, @AlexanderSher!  A couple of nits, but almost ready to approve.

File: test/AutoRest.TestServerLowLevel.Tests/dpg-customization.cs (hunk 4 above)
Comment (AlexanderSher) [reply test/AutoRest.TestServerLowLevel.Tests/dpg-customization.cs:91]: Ok

File: test/AutoRest.TestServerLowLevel.Tests/dpg-customization.cs (hunk 5 above)
Comment (AlexanderSher) [reply test/AutoRest.TestServerLowLevel.Tests/dpg-customization.cs:89]: No, as I said, the only indirect way to check that is that previous scope is closed before `WaitForCompletionAsync` is closed.

Comment (annelo-msft) [reply test/AutoRest.TestServerLowLevel.Tests/dpg-customization.cs:89]: @JoshLove-msft @christothes, FYI -- this is one of the gaps in our tests we were discussing yesterday.  Looks like we aren't able to validate a hierarchical relationship using our test infra.

File: test/AutoRest.TestServerLowLevel.Tests/dpg-customization.cs [hunk 7]
```diff
@@ -74,6 +74,61 @@ await foreach (Page<BinaryData> page in allPages.AsPages())
...
+        [Test]
+        public Task HandwrittenModelLro() => Test(async (host) =>
+        {
+            using var diagnosticListener = new ClientDiagnosticListener("dpg_customization_LowLevel", asyncLocal: true);
+            CollectionAssert.IsEmpty(diagnosticListener.Scopes);
+
+            Operation<Product> lro = await new DPGClient(Key, host).LroValueAsync(WaitUntil.Started, "model");
+            Assert.AreEqual(1, diagnosticListener.Scopes.Count);
+            Assert.AreEqual("DPGClient.LroValue", diagnosticListener.Scopes[0].Name);
+            Assert.IsTrue(diagnosticListener.Scopes[0].IsCompleted);
```
Comment (JoshLove-msft) [inline test/AutoRest.TestServerLowLevel.Tests/dpg-customization.cs:86]: You can use [AssertAndRemoveScope](https://github.com/Azure/azure-sdk-for-net/blob/72fc7b6182f15d46a8afb529af2f5fb6730c4133/sdk/core/Azure.Core.TestFramework/src/ClientDiagnosticListener.cs#L201) which will throw if the scope is not completed and allow you to specify a scope name and attributes that the scope should contain.

**Review by annelo-msft — APPROVED** [review]

File: test/TestServerProjectsLowLevel/dpg-customization/Customization/DPGClient.cs (hunk 2 above)
Comment (AlexanderSher) [reply test/TestServerProjectsLowLevel/dpg-customization/Customization/DPGClient.cs]: Fixed

File: test/AutoRest.TestServerLowLevel.Tests/dpg-customization.cs (hunk 7 above)
Comment (AlexanderSher) [reply test/AutoRest.TestServerLowLevel.Tests/dpg-customization.cs]: Fixed

**Comment by pshao25** [comment]
RawLRO not covered here?

**Comment by annelo-msft** [comment]
> RawLRO not covered here?

@pshao25, what is RawLRO?

**Comment by pshao25** [comment]
> > RawLRO not covered here?
> 
> @pshao25, what is RawLRO?

I have added at https://github.com/Azure/autorest.csharp/pull/2211

**Pull request merged**